### ISSUE TITLE
[OpenEnv] multi step training issue in GRPO trainer

### ISSUE THREAD

**Issue by akshay-gup** [huggingface/trl#4543]
### Reproduction

Hi, I have a question/concern about using `GRPOTrainer` for multi-step agent training in **vLLM server mode**.

In `grpo_trainer.py`, inside the server-mode generation path, there is this line:

```python
# At this point, we only get 1 copy of each prompt, so we need to repeat them num_generations times
all_prompt_ids = [ids for ids in all_prompt_ids for _ in range(self.num_generations)]
```

This works correctly when:

* using normal one-shot generation, or
* using a `rollout_func` that produces **single-step** completions (one completion per generation).

However, for **multi-step agents**, the trajectory is produced over multiple turns, and each turn has a different prefix (because environment/tool responses get appended to the prompt at every step).
In that case, **each rollout has its own sequence of step-wise prefixes**, which leads to different `prompt_ids` per rollout.

Because server-mode GRPO duplicates the original prompt and pairs it with all G completions, we effectively lose all per-step prefixes. This breaks the assumptions behind importance sampling and causes the IS ratio to be computed against the wrong behavior policy.

In other words:

* multi-step trajectories require **different `prompt_ids` for each rollout**, because the agent conditions on different environment states at each decision point,
* but the trainer forces **one shared prompt per dataset example**, duplicated `num_generations` times,
* so step-wise importance sampling becomes mathematically incorrect in server mode.

Is multi-step agent training intended to be supported in server mode?
Or should the trainer be extended to accept per-rollout (or per-step) prefixes instead of forcing all rollouts to share the same `prompt_ids`?

Thanks!

### System Info

- Platform: Linux-5.15.0-152-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- TRL version: 0.23.1
- PyTorch version: 2.8.0
- accelerator(s): NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000, NVIDIA RTX A6000
- Transformers version: 4.56.2
- Accelerate version: 1.10.1
- Accelerate config: not found
- Datasets version: 4.2.0
- HF Hub version: 0.35.2
- bitsandbytes version: 0.48.1
- DeepSpeed version: 0.18.0
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.109.1
- PEFT version: 0.17.1
- vLLM version: 0.11.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by MohamedAliRashad** [comment]
I don't understand how `prompt_ids` change with every rollout ?
if the prompt is the user query, shouldn't it be fixed whatever the multistep completion looks like ?

**Comment by gmanlan** [comment]
I’ve also encountered this issue/limitation, but wanted to provide a different perspective that may clarify the issue and/or severity. I’m using vLLM in colocate mode, but I don’t think this is a relevant detail.

GRPOTrainer supports two approaches: environment_factory (black box) and rollout_func (white box).

In my opinion, the main reason why GRPOTrainer with custom rollout_func (or environment_factory) exists is to support Multi-turn GRPO. If we wanted to run single-turn/single-generation Q&A kind of problems, we would have used the [basic “hello world" GRPOTrainer interface](https://huggingface.co/docs/trl/grpo_trainer#quick-start) that doesn’t require the definition of an environment or rollout function.

When using the rollout_func approach, we must implement our own agent loop and use generate_rollout_completions() to obtain logprobs, prompt_ids, completion_ids, etc. This data is then returned by the rollout_func so that GRPOTrainer can properly do training (compute advantages, apply weights, and so on). The expected result format of rollout_func is:

{
“prompt_ids": list[int],
“completion_ids”: list[int],
“logprobs”: list[float],
… rewards, etc.
}

Considering that multi-turn RL is typically associated with agentic use (tool calling, etc.), the closest example notebook I have seen is: https://github.com/huggingface/trl/blob/main/examples/notebooks/grpo_functiongemma_browsergym_openenv.ipynb

This example notebook specifically defines a rollout_once function, and if you look at how the prompt_ids are collected after calling generate_rollout_completions, it’s using prompt_ids.extend(…). Extend literally adds all the prompt_ids of the current turn to all the prompt_ids of the previous turn, which technically duplicates them N times (N turns). Maybe the notebook example doesn’t suffer from this problem because it behaves as a “single-turn” procedure and doesn’t accumulate a message history necessarily? (not in the traditional sense anyways).

In multi-turn, this is a typical example (U = user, A = assistant, TR = tool response):

Turn 0 = prompt [SYS, U1] —> generation [A1, TR1]
Turn 1 = prompt [SYS, U1, A1, TR1] —> generation [U2]
Turn 2 = prompt [SYS, U1, A1, TR1, U2] —> generation [A2, TOOL_RESPONSE]
and so on…

Consequently, the prompt_ids for turn 0 are different from turn 1 and so on. As the message history accumulates, the input tokens grow.

What I would have expected is for the trainer to be able to somehow receive a list of “turns”, with each “turn" containing its respective list of prompt_ids, completion_ids, logprobs. Something like:

[
turn_0: {prompt_ids: [], completion_ids: [], logprobs: []},
turn_1: {prompt_ids: [], completion_ids: [], logprobs: []},
turn_2: {prompt_ids: [], completion_ids: [], logprobs: []}
]

At the moment I believe GRPOTrainer is simply assuming it’s always a single-turn case (so the single prompt_ids -flattened- list corresponds only to turn 0).

I do notice an optional env_mask extra field that would be used to properly mask non-generated content (if tools = False when calling the trainer). I’ve tried building a custom snippet inside my rollout function that would do some creative work around adding turn_1 onwards extra prompt_ids as part of the generation but then use the mask to assign a loss = 0 on the (prompt) tokens I don’t want to compute and loss = 1 on the actual generation, but this sounds cumbersome and I’m not entirely sure is fully compatible with the GRPOTrainer internals (currently debugging).

1) Am I completely wrong and I missed something obvious here?
2) Do we have any genuine multi-turn/tool-calling example that uses the rollout function (not the environment_factory approach that does agent looping automagically) and that has been proven to work?

**Comment by zhzhang** [comment]
@gmanlan consider using https://unsloth.ai/docs/get-started/reinforcement-learning-rl-guide/training-ai-agents-with-rl for your use case

**Comment by albertvillanova** [comment]
Hey @akshay-gup, sorry for the slow reply here — we're catching up on the backlog. 

Good catch: the line you quoted (`all_prompt_ids = [ids for ids in all_prompt_ids for _ in range(self.num_generations)]`) was part of the old server-mode path, which deduped prompts, called `rollout_func` on the unique set, then re-expanded — discarding any per-rollout `prompt_ids` your rollout returned. 

That whole path was removed in the vLLM generation refactor (#4700). On current `main` the `rollout_func` receives the fully-repeated prompt list and its returned per-rollout `prompt_ids`/`completion_ids`/`logprobs` are used verbatim — no forced duplication — so different environment prefixes per rollout are now supported: https://github.com/huggingface/trl/blob/378fe7c04fb2f60d42b99803ab99de4c43653ef3/trl/trainer/grpo_trainer.py#L1849-L1866

On the deeper point: a multi-step trajectory isn't represented as separate `prompt_ids` per turn. It's one flattened rollout — a single prompt prefix plus a completion that interleaves model and environment tokens — with an `env_mask` that zeroes the loss on the environment tokens (https://github.com/huggingface/trl/blob/378fe7c04fb2f60d42b99803ab99de4c43653ef3/trl/trainer/grpo_trainer.py#L1911-L1913). Because you return the actual per-token sampling `logprobs`, the per-token IS ratio is still computed against the true behavior policy — the mask just excludes non-model tokens from the loss. `environment_factory` builds that mask for you automatically; the multi-turn design is documented in the OpenEnv guide: https://huggingface.co/docs/trl/main/openenv

Closing as fixed, but happy to reopen if you still see per-rollout prompts being collapsed on a current TRL version.
